README.md
<div align="center">
  <img src="https://raw.githubusercontent.com/mirkobrombin/Amusiz/master/data/pm.mirko.Amusiz.svg" width="64">
  <h1 align="center">Amusiz</h1>
  <p align="center">An Amazon Music client for Linux (unpretentious)</p>
</div>

<br/>

<div align="center">
  <a href="https://www.codefactor.io/repository/github/mirkobrombin/amusiz">
    <img src="https://www.codefactor.io/repository/github/mirkobrombin/amusiz/badge" alt="CodeFactor" />
  </a>
  <a href="https://github.com/mirkobrombin/Amusiz/blob/master/LICENSE">
    <img src="https://img.shields.io/badge/License-GPL--3.0-blue.svg">
  </a>
  <a href="https://github.com/mirkobrombin/Amusiz/actions">
    <img src="https://github.com/mirkobrombin/Amusiz/workflows/Build%20release%20packages/badge.svg">
  </a>
  <a href="https://aur.archlinux.org/packages/amusiz/">
    <img alt="AUR version" src="https://img.shields.io/aur/version/amusiz">
  </a>
</div>

<br />

<div align="center">
    <img  src="https://raw.githubusercontent.com/mirkobrombin/Amusiz/main/data/screenshot.png">
</div>

## ↗️ Install
You can install Amusiz in multiple ways, choose your favorite.

### 🚀 AppImage
[Here](https://github.com/mirkobrombin/Amusiz/releases) you can find the latest AppImage build.

### AUR
There is the official package of Amusiz on AUR for ArchLinux and derivatives.

<a href='https://aur.archlinux.org/packages/amusiz/'>
  <img alt="AUR version" src="https://img.shields.io/aur/version/amusiz?style=for-the-badge">
</a>

### Snap
[![Get it from the Snap Store](https://snapcraft.io/static/images/badges/en/snap-store-black.svg)](https://snapcraft.io/amusiz)

### Flatpak build
```bash
flatpak-builder --repo=amusiz --force-clean --user build-dir pm.mirko.Amusiz.yml
flatpak remote-add --user amusiz amusiz --no-gpg-verify
flatpak install --user amusiz pm.mirko.Amusiz
```

### Build dependencies
- meson
- ninja
- libhandy
- libwebkit2gtk-4.0
- gir1.2-gtk-3.0
- gir1.2-handy-1
- gir1.2-notify-0.7
- gir1.2-webkit2-4.0
- libgstreamer1.0-0
- gstreamer1.0-plugins-base
- gstreamer1.0-plugins-good
- gstreamer1.0-libav
- gstreamer1.0-alsa
- gstreamer1.0-pulseaudio

### Know issues
**Amusiz ask me to re-login each time**  
Choose the correct language from the preferences.

### 🛠️ Build
```bash
meson build
cd build
ninja install
```


debian/changelog
pm.mirko.amusiz (1.3.3) precise; urgency=low

  * update flatpak manifest

 -- Mirko Brombin <send@mirko.pm>  Thu, 09 Jul 2021 14:31:00 +0000


src/about.py
# about.py
#
# Copyright 2021 Mirko Brombin
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

from gi.repository import Gtk

@Gtk.Template(resource_path='/pm/mirko/Amusiz/about.ui')
class AmusizAboutDialog(Gtk.AboutDialog):
    __gtype_name__ = 'AmusizAboutDialog'

    def __init__(self, window, **kwargs):
        super().__init__(**kwargs)
        self.set_transient_for(window)



src/window.py
# window.py
#
# Copyright 2021 Mirko Brombin
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import os
import webbrowser
from gi.repository import Gtk, Gdk, Gio, Handy, WebKit2
from pathlib import Path
from . import webview
from . import about
from . import preferences


@Gtk.Template(resource_path='/pm/mirko/Amusiz/window.ui')
class AmusizWindow(Handy.ApplicationWindow):
    __gtype_name__ = 'AmusizWindow'

    btn_refresh = Gtk.Template.Child()
    entry_search = Gtk.Template.Child()
    scroll_window = Gtk.Template.Child()
    btn_preferences = Gtk.Template.Child()
    btn_about = Gtk.Template.Child()
    btn_help = Gtk.Template.Child()

    default_settings = Gtk.Settings.get_default()
    settings = Gio.Settings.new("pm.mirko.Amusiz")

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        Handy.init()

        self.webview = webview.AmusizWebView(self)

        '''Prefer dark theme'''
        self.default_settings.set_property(
            "gtk-application-prefer-dark-theme",
            True
        )

        '''Signals'''
        self.btn_refresh.connect('pressed', self.webview.on_refresh)
        self.btn_preferences.connect('pressed', self.show_preferences)
        self.btn_about.connect('pressed', self.show_about)
        self.btn_help.connect('pressed', self.show_help)
        self.entry_search.connect('key-release-event', self.webview.on_search)
        self.entry_search.connect('focus-in-event', self.webview.hit_element, "#navbarSearchInput")
        self.entry_search.connect('icon-release', self.webview.clear_search)
        self.settings.connect('changed', self.webview.start, "lang")
        if self.settings.get_boolean("search"):
            self.entry_search.connect('activate', self.webview.perform_adv_search)

        '''Disclaimer'''
        if self.settings.get_boolean("disclaimer"):
            disclaimer_window = AmusizDisclaimer(self)
            disclaimer_window.present()

        '''Show widgets'''
        self.scroll_window.add(self.webview.webview)
        self.show_all()

    def show_preferences(self, widget):
        p = preferences.AmusizPreferences(self)
        p.present()

    def show_about(self, widget):
        about.AmusizAboutDialog(self).show_all()

    @staticmethod
    def show_help(widget):
        webbrowser.open("https://github.com/mirkobrombin/Amusiz")

@Gtk.Template(resource_path='/pm/mirko/Amusiz/disclaimer.ui')
class AmusizDisclaimer(Handy.Window):
    __gtype_name__ = 'AmusizDisclaimer'

    def __init__(self, window, **kwargs):
        super().__init__(**kwargs)
        self.set_transient_for(window)
        self.window = window

        '''Signals'''
        self.connect('delete-event', self.close)

    def close(self, widget, data):
        self.window.settings.set_boolean("disclaimer", False)


src/scripts.js
function DecreaseVolume() {
    var volBar = document.getElementById('volume-range')
    if (volBar === null) {
        document.getElementById('volume-button').click()
        setTimeout(() => {
            DecreaseVolume()
        }, 300)
        return
    }
    var currentVolume = parseFloat(volBar.value)
    var newVolume = currentVolume - 0.05
    if(newVolume < 0) newVolume = 0.0
    volBar.value = newVolume.toString()
    if ("createEvent" in document) {
        var evt = document.createEvent("HTMLEvents");
        evt.initEvent("change", false, true);
        volBar.dispatchEvent(evt);
    }
    else
        volBar.fireEvent("onchange");
}

function IncreaseVolume() {
    var volBar = document.getElementById('volume-range')
    if (volBar === null) {
        document.getElementById('volume-button').click()
        setTimeout(() => {
            DecreaseVolume()
        }, 300)
        return
    }
    var currentVolume = parseFloat(volBar.value)
    var newVolume = currentVolume + 0.05
    if(newVolume > 1) newVolume = 1.0
    volBar.value = newVolume.toString()
    if ("createEvent" in document) {
        var evt = document.createEvent("HTMLEvents");
        evt.initEvent("change", false, true);
        volBar.dispatchEvent(evt);
    }
    else
        volBar.fireEvent("onchange");
}

function fixVolumeBar() {
    var volBar = document.getElementById('volume-range')
    if (volBar === null) return

    volBar.setAttribute('step', '0.05')
}
setInterval(() => fixVolumeBar(), 300)


/*var ready = function () {

  document.getElementById("accountSetting").click();

	if (document.body && document.querySelector("#contextMenuOverlay[aria-label='menu']")) {
    setTimeout(() => {
        user = document.querySelector("#contextMenuOverlay[aria-label='menu'] > div > music-context-menu-header").getAttribute("primary-text");
        alert(user)
    }, 300)
		return;
	}
	window.requestAnimationFrame(ready);
};

window.requestAnimationFrame(ready);*/


src/webview.py
# webview.py
#
# Copyright 2021 Mirko Brombin
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import webbrowser
import urllib.parse
from gi.repository import Gtk, Gdk, Gio, Handy, WebKit2
from .globals import amazon_uris, cookies_path


class AmusizWebSettings(WebKit2.Settings):

    hw_accell = WebKit2.HardwareAccelerationPolicy.NEVER

    def __init__(self, **kwargs):
        super().__init__(**kwargs)

        self.set_hardware_acceleration_policy(self.hw_accell)
        self.set_enable_back_forward_navigation_gestures(True)
        self.set_enable_dns_prefetching(True)
        self.set_enable_media_capabilities(True)
        # self.set_enable_smooth_scrolling(True)
        self.set_enable_developer_extras(True)


class AmusizContentManager(WebKit2.UserContentManager):

    def __init__(self, **kwargs):
        super().__init__(**kwargs)

        '''CSS tricks'''
        style = Gio.resources_lookup_data("/pm/mirko/Amusiz/inject.css", 0)
        self.add_style_sheet(
            WebKit2.UserStyleSheet(
                str(style.get_data(), "utf-8"),
                WebKit2.UserContentInjectedFrames.TOP_FRAME,
                WebKit2.UserStyleLevel.USER,
                None, None
            )
        )


class AmusizWebView():

    context = WebKit2.WebContext.get_default()
    manager = AmusizContentManager()
    settings = AmusizWebSettings()
    policies = WebKit2.WebsitePolicies(autoplay=WebKit2.AutoplayPolicy.ALLOW)
    webview = WebKit2.WebView(
        settings=settings,
        user_content_manager=manager,
        website_policies=policies
    )
    cookies = context.get_cookie_manager()
    cookies.set_accept_policy(WebKit2.CookieAcceptPolicy.ALWAYS)

    def __init__(self, window, **kwargs):
        super().__init__(**kwargs)

        self.window = window
        self.start()

        '''Signals'''
        self.webview.connect('load-changed', self.on_change)
        self.webview.connect('load-failed-with-tls-errors', self.on_tls_errors)

        '''Webview'''
        self.webview.set_background_color(Gdk.RGBA(0.05, 0.05, 0.05, 1.0))

        '''Cookies'''
        self.cookies.set_persistent_storage(
            cookies_path,
            WebKit2.CookiePersistentStorage.SQLITE
        )

    '''Webview methods'''

    def start(self, settings=False, key=None, user_data=None):
        lang = self.window.settings.get_string("lang")
        self.amazon_uri = amazon_uris[lang]
        self.webview.load_uri(self.amazon_uri)

    def hit_element(self, widget=None, user_data=None, element="body"):
        script = f"document.querySelector('{element}').click();"
        self.webview.run_javascript(script)

    def open_browser(self, widget=None, url="about:blank"):
        webbrowser.open(url)

    def open_url(self, widget=None, url="about:blank"):
        self.webview.load_uri(url)

    def on_tls_errors(self, widget, failing_uri, certificate, errors):
        host = urllib.parse.urlparse(failing_uri).netloc
        ctx = self.webview.get_context()
        ctx.allow_tls_certificate_for_host(certificate, host)

    def on_refresh(self, widget=None):
        self.webview.reload()

    def clear_search(self, widget, icon_pos, event, user_data=None):
        self.on_search(widget, Gdk.KEY_Escape)

    def perform_adv_search(self, widget=None, data=None):
        terms = widget.get_text()
        url = f"{self.amazon_uri}/search/{terms}"
        self.webview.load_uri(url)

    def on_search(self, widget=None, event=None, data=None):
        if event == Gdk.KEY_Escape or event.keyval == Gdk.KEY_Escape:
                widget.set_text("")

        terms = widget.get_text()
        script = f"""
        document.getElementById("navbarSearchInput").value ="{terms}";
        var event = document.createEvent('Event');
        event.initEvent('input', true, true);
        document.getElementById("navbarSearchInput").dispatchEvent(event);
        """
        self.webview.run_javascript(script)

    def on_change(self, web_view, load_event):
        scripts = Gio.resources_lookup_data("/pm/mirko/Amusiz/scripts.js", 0)
        self.webview.run_javascript(str(scripts.get_data(), "utf-8"))




src/preferences.py
# preferences.py
#
# Copyright 2020 brombinmirko <send@mirko.pm>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

from gi.repository import Gtk, GLib, Handy
from .globals import amazon_uris

@Gtk.Template(resource_path='/pm/mirko/Amusiz/preferences.ui')
class AmusizPreferences(Handy.PreferencesWindow):
    __gtype_name__ = 'AmusizPreferences'

    '''Get widgets from template'''
    combo_lang = Gtk.Template.Child()
    switch_advanced_search = Gtk.Template.Child()

    def __init__(self, window, **kwargs):
        super().__init__(**kwargs)
        self.set_transient_for(window)

        '''Common variables'''
        self.window = window
        self.settings = window.settings
        self.default_settings = window.default_settings

        '''Set widgets status from user settings'''
        self.combo_lang.set_active_id(self.settings.get_string("lang"))
        self.switch_advanced_search.set_active(self.settings.get_boolean("search"))

        '''Signal connections'''
        self.combo_lang.connect('changed', self.change_language)
        self.switch_advanced_search.connect('state-set', self.toggle_advanced_search)

    def change_language(self, widget):
        res = widget.get_active_id()
        self.settings.set_string("lang", res)

    def toggle_advanced_search(self, widget, state):
        self.settings.set_boolean("search", state)
        if state:
            self.window.entry_search.connect(
                'activate',
                self.window.webview.perform_adv_search
            )
        


src/main.py
# main.py
#
# Copyright 2021 Mirko Brombin
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import sys
import gi
import webbrowser

gi.require_version('Gtk', '3.0')
gi.require_version('Handy', '1')
gi.require_version('WebKit2', '4.0')

from gi.repository import Gtk, Gdk, Gio, Handy

from .window import AmusizWindow


class Application(Gtk.Application):
    def __init__(self):
        super().__init__(application_id='pm.mirko.Amusiz',
                         flags=Gio.ApplicationFlags.FLAGS_NONE)

    def do_startup(self):
        Gtk.Application.do_startup(self)
        self.set_actions()

    def do_activate(self):
        '''Load custom CSS'''
        data_bytes = Gio.resources_lookup_data(
            "/pm/mirko/Amusiz/style.css", 0)
        provider = Gtk.CssProvider()
        provider.load_from_data(data_bytes.get_data())
        Gtk.StyleContext.add_provider_for_screen(Gdk.Screen.get_default(),
                                                 provider,
                                                 Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION)

        win = self.props.active_window
        if not win:
            win = AmusizWindow(application=self)

        self.win = win
        win.present()


    def quit(self, action=None, param=None):
        '''Quit application [CTRL+Q]'''
        self.win.destroy()

    @staticmethod
    def help(action, param):
        '''Open Help URL [F1]'''
        webbrowser.open_new_tab("https://github.com/mirkobrombin/Amusiz")

    def refresh(self, action, param):
        '''Refresh [CTRL+R]'''
        self.win.on_refresh()

    def set_actions(self):
        '''Register window actions'''
        action_entries = [
            ("quit", self.quit, ("app.quit", ["<Ctrl>Q"])),
            ("help", self.help, ("app.help", ["F1"])),
            ("refresh", self.refresh, ("app.refresh", ["<Ctrl>R"]))
        ]

        for action, callback, accel in action_entries:
            simple_action = Gio.SimpleAction.new(action, None)
            simple_action.connect('activate', callback)
            self.add_action(simple_action)
            if accel is not None:
                self.set_accels_for_action(*accel)


def main(version):
    app = Application()
    return app.run(sys.argv)
